sw/device/lib/crypto/impl/mac.c

```diff
@@ -9,23 +9,109 @@
 #include "sw/device/lib/crypto/impl/keyblob.h"
-#include "sw/device/lib/crypto/impl/sha2/sha256.h"
-#include "sw/device/lib/crypto/impl/sha2/sha512.h"
 #include "sw/device/lib/crypto/impl/status.h"
 #include "sw/device/lib/crypto/include/hash.h"
 
 // Module ID for status codes.
 #define MODULE_ID MAKE_MODULE_ID('m', 'a', 'c')
 
+/**
+ * Ensure that the hash context is large enough for HMAC driver struct.
+ */
+static_assert(
+    sizeof(otcrypto_hmac_context_t) >= sizeof(hmac_ctx_t),
+    "`otcrypto_hash_context_t` must be big enough to hold `hmac_ctx_t`.");
+
+/**
+ * Ensure that HMAC driver struct is suitable for `hardened_memcpy()`.
+ */
+static_assert(sizeof(hmac_ctx_t) % sizeof(uint32_t) == 0,
+              "Size of `hmac_ctx_t` must be a multiple of the word size for "
+              "`hardened_memcpy()`");
+
+/**
+ * Save the internal HMAC driver context to a generic Hmac context.
+ *
+ * @param[out] ctx Generic hash context to copy to.
+ * @param hmac_ctx The internal context object from HMAC driver.
+ */
+static void hmac_ctx_save(otcrypto_hmac_context_t *restrict ctx,
+                          const hmac_ctx_t *restrict hmac_ctx) {
+  hardened_memcpy(ctx->data, (uint32_t *)hmac_ctx,
+                  sizeof(hmac_ctx_t) / sizeof(uint32_t));
+}
+
+/**
+ * Restore an internal HMAC driver context from a generic Hmac context.
+ *
+ * @param ctx Generic hash context to restore from.
+ * @param[out] hmac_ctx Destination HMAC driver context object.
+ */
+static void hmac_ctx_restore(const otcrypto_hmac_context_t *restrict ctx,
+                             hmac_ctx_t *restrict hmac_ctx) {
+  hardened_memcpy((uint32_t *)hmac_ctx, ctx->data,
+                  sizeof(hmac_ctx_t) / sizeof(uint32_t));
+}
+
+static status_t get_hmac_mode(otcrypto_key_mode_t key_mode,
+                              hmac_mode_t *hmac_mode) {
+  switch (key_mode) {
+    case kOtcryptoKeyModeHmacSha256:
+      *hmac_mode = kHmacModeHmac256;
+      break;
+    case kOtcryptoKeyModeHmacSha384:
+      *hmac_mode = kHmacModeHmac384;
+      break;
+    case kOtcryptoKeyModeHmacSha512:
+      *hmac_mode = kHmacModeHmac512;
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
+  }
+  return OTCRYPTO_OK;
+}
+
 OT_WARN_UNUSED_RESULT
 otcrypto_status_t otcrypto_hmac(const otcrypto_blinded_key_t *key,
                                 otcrypto_const_byte_buf_t input_message,
                                 otcrypto_word32_buf_t tag) {
-  // Compute HMAC using the streaming API.
-  otcrypto_hmac_context_t ctx;
-  HARDENED_TRY(otcrypto_hmac_init(&ctx, key));
-  HARDENED_TRY(otcrypto_hmac_update(&ctx, input_message));
-  return otcrypto_hmac_final(&ctx, tag);
+  // Validate key struct.
+  if (key == NULL || key->keyblob == NULL || tag.data == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+  if (key->config.hw_backed != kHardenedBoolFalse) {
+    // TODO(#15590): Add support for sideloaded keys via a custom OTBN program.
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
```

Comment:
I assume this is no longer valid as we are now using the `hmac` block.

Reply:
Removing sideload and masked implementation TODOs and adding the following comment line:
```
// HMAC HWIP is not masked and it does not have sideload support, so the
// following conditions return not implemented.
```